### DISCUSSION TITLE
How to forward declare function that referenced a struct?

### DISCUSSION THREAD

**Discussion by codehz** [edubart/nelua-lang#109] [Q&A] [answered]
example:

```lua
local F = @function(): S
local S = @record{ f: F }
```

got error: `undeclared symbol 'S'`

**Comment by edubart** [comment, marked as answer]
You can only do this using `<forwarddecl>` annotation, here is an example:

```lua
local S <forwarddecl> = @record{} -- declare S with incomplete fields
local F = @function(): S -- use S
S = @record{ f: F } -- define S fields
```

Another option is to use `S` inside its definition:

```lua
local S = @record{ f: function(): S }
```

**Reply by edubart** [reply]
The requirement of using pointers to incomplete types was relaxed for function references since commit https://github.com/edubart/nelua-lang/commit/14acb9133b2f21e3d84189bcb975694583c38740 , I've edited the answer accordingly.
